cpp/src/strings/split/split.cu

```diff
@@ -385,6 +386,348 @@ struct whitespace_rsplit_tokenizer_fn
+#if 1
+template <typename T>
+__device__ inline T round_up_pow2(T number_to_round, T modulus) {
+    return (number_to_round + (modulus - 1)) & -modulus;
+}
+
+// align all column size allocations to this boundary so that all output column buffers
+// start at that alignment.
+static constexpr size_type split_align = 64;
+#endif
+
+/**
+ * @brief Compute the number of tokens, the total byte sizes of the tokens, and
+ * required memory size for the `idx'th` string element of `d_strings`.
+ */
+template <bool forward>
+struct token_reader_fn {
+  column_device_view const d_strings;  // strings to split
+  string_view const d_delimiter;  // delimiter for split
+  size_type const max_tokens = std::numeric_limits<size_type>::max();
+  bool const has_validity = false;
+
+  // returns a tuple of token count, sum of token sizes in bytes, and required
+  // memory block size
+  __device__ thrust::tuple<size_type, size_type, size_type>
+  operator()(size_type idx) const {
+    if (has_validity && d_strings.is_null(idx)) {
+      return thrust::make_tuple<size_type, size_type, size_type>(0, 0, 0);
+    }
+
+    auto d_str = d_strings.element<string_view>(idx);
+    size_type token_count = 0;
+    size_type token_size_sum = 0;
+    size_type start_pos = 0;  // updates only if forward is true
+    auto end_pos = d_str.length();  // updates only if forward is false
+    while (token_count < max_tokens - 1) {
+      auto delimiter_pos =
+        forward ? d_str.find(d_delimiter, start_pos) :
+                  d_str.rfind(d_delimiter, start_pos, end_pos);
+      if (delimiter_pos != -1) {
+        auto token_char_bytes =
+          forward ?
+            d_str.byte_offset(delimiter_pos) - d_str.byte_offset(start_pos) :
+            d_str.byte_offset(end_pos) -
+              d_str.byte_offset(delimiter_pos + d_delimiter.length());
+        token_count++;
+        token_size_sum += token_char_bytes;
+        if (forward) {
+          start_pos = delimiter_pos + d_delimiter.length();
+        }
+        else {
+          end_pos = delimiter_pos;
+        }
+      }
+      else {
+        break;
+      }
+    }
+    token_count++;
+    token_size_sum +=
+      forward ? d_str.byte_offset(end_pos) - d_str.byte_offset(start_pos) :
+                d_str.byte_offset(end_pos);
+
+    auto memory_size =
+      round_up_pow2(token_size_sum, split_align) +
+      round_up_pow2(
+        (token_count + 1) * static_cast<size_type>(sizeof(size_type)),
+        split_align);
+
+    return thrust::make_tuple<size_type, size_type, size_type>(
+      token_count, token_size_sum, memory_size);
+  }
+};
+
+/**
+ * @brief Copy the tokens from the `idx'th` string element of `d_strings` to
+ * the contiguous memory buffer.
+ */
+template <bool forward>
+struct token_copier_fn {
+  column_device_view const d_strings;  // strings to split
+  string_view const d_delimiter;  // delimiter for split
+  bool const has_validity = false;
+
+  __device__ void
+  operator()(
```

Comment:
Consider refactoring `operator()` into separate calls to several `__device__` member functions for readability.